### PR TITLE
Add POC document

### LINKED ISSUE #23: TRACKER: Checklist for the POC
Expand our `generalized_window.rst` proposal document as POC showcasing:

- [x] Performance parity of our rolling mean implementation (https://github.com/twosigma/pandas/issues/17)
- [x] `GroupbyIndexer` and/or `ExpandingIndexer` example
- [x] Performance of `rolling.apply`

### PR DESCRIPTION
**Pull request by mroeschke**
- [x] closes #23

### CODE REVIEW

**Comment by mroeschke** [comment]
Quick performance considerations this doesn't include:

* The current `VariableWindowIndexer` and `FixedWindowIndexer` are still using `jitclass` and the performance hit is especially bad for offset rolling. I will put up another PR for this change as local testing showed I can get another 2x performance boost for rolling mean.
* The current roll_mean cython implementation notes that there's a separate path for fixed rolling mean which is faster but it will make the code more complex. Not sure if it's worth increasing code complexity, but it will depend on the speed up achieved.
https://github.com/pandas-dev/pandas/blob/9ab347808e8fd170ef68276c3edb8246b5d30cb4/pandas/_libs/window.pyx#L584

File: doc/source/development/rolling_operations_with_numba.rst [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+#. Compute ``rolling.apply``
+#. Calculate offset-based or numeric-based window boundaries.
+
+The benefits of using ``Numba`` include:
+
+#. Performance parity or improvement over ``Cython``
+#. Eliminate shipping C-extensions
+#. Ease of debugging
+
+Additionally this POC includes an API that allows users to create custom window boundary calculations.
```
Comment (DiegoAlbertoTorres) [inline doc/source/development/rolling_operations_with_numba.rst]: Let's add a section where we give a concrete proposal for shipping these changes: in Pandas 1.0 (and add link to the roadmap), with Numba as a hard dependency, etc.

Comment (DiegoAlbertoTorres) [reply doc/source/development/rolling_operations_with_numba.rst]: Also mention docs and todos

File: doc/source/development/rolling_operations_with_numba.rst [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+Timings on master:
+
+.. code-block:: ipython
+
+   In [1]: %load_ext memory_profiler
+
+   In [2]: pd.__version__
+   Out[2]: '0.26.0.dev0+514.g24b1dd128'
+
+   In [3]: n = 1_000_000
```
Comment (DiegoAlbertoTorres) [inline doc/source/development/rolling_operations_with_numba.rst:165]: Let's try these with 10M, 100M and 1B

File: doc/source/development/rolling_operations_with_numba.rst [hunk 3]
```diff
@@ -0,0 +1,173 @@
+Rolling Operations with Numba
+=============================
+
+Summary
+-------
+
```
Comment (jreback) [inline doc/source/development/rolling_operations_with_numba.rst:6]: let’s add a Future section
- use jitclasses (mention issues and resolution: https://github.com/numba/numba/issues/4522#issuecomment-537872456)

can include what we want short term (implement rest of the aggregations, expanding, ewma)

and longer term; groupby & groupby + rolling

File: doc/source/development/rolling_operations_with_numba.rst (hunk 2 above)
Comment (mroeschke) [reply doc/source/development/rolling_operations_with_numba.rst:165]: For memory, which we were most curious about, at 1B the Numba implimentation is finally starting to edge out Cython. Does that align with your experience?

```
# POC
In [1]: %load_ext memory_profiler

In [2]: n = 1_000_000_000

In [3]: roll_fixed = pd.Series(range(n)).rolling(10)

In [4]: %memit roll_fixed.apply(lambda x: np.sum(x) + 5, raw=True)
peak memory: 11747.92 MiB, increment: 4056.36 MiB

# master
In [1]: %load_ext memory_profiler

In [2]: n = 1_000_000_000

In [3]: roll_fixed = pd.Series(range(n)).rolling(10)

In [4]: %memit roll_fixed.apply(lambda x: np.sum(x) + 5, raw=True)
peak memory: 12581.71 MiB, increment: 4860.52 MiB
```

File: doc/source/development/rolling_operations_with_numba.rst [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+The functional implementation mimics the current ``Cython`` implementation; however, now that
+the calculation of the window boundaries is separate from the aggregation function, we are able to
+expose a new API to users that allows them to specify how to calculate the rolling window boundaries.
+
+New Custom Window Indexer API
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Currently, window bounds are calculated automatically based on the ``DataFrame`` or ``Series`` index
+when a user passes an integer or offset in the rolling API (e.g. ``df.rolling(2)``). This POC also introduces
+a ``BaseIndexer`` class is available from ``pandas/core/window/indexers.py`` for users to subclass
```
Comment (DiegoAlbertoTorres) [inline doc/source/development/rolling_operations_with_numba.rst]: s/is available from/in

File: doc/source/development/rolling_operations_with_numba.rst [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+to create a custom routine to calculate window boundaries. Users will need to specify a
+``get_window_bounds`` function to calculate window boundaries.
+
+Below is an example of creating an ``ExpandingIndexer`` computes an expanding window similar to
```
Comment (DiegoAlbertoTorres) [inline doc/source/development/rolling_operations_with_numba.rst]: s/computes/that computes

File: doc/source/development/rolling_operations_with_numba.rst [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+   1    0.5
+   2    1.0
+   3    1.5
+   4    2.0
+   dtype: float64
+
+Performance
+-----------
+
+Below is table comparing the current performance difference between the Numba and Cython implementations
```
Comment (DiegoAlbertoTorres) [inline doc/source/development/rolling_operations_with_numba.rst]: s/is/is a

File: doc/source/development/rolling_operations_with_numba.rst (hunk 2 above)
Comment (DiegoAlbertoTorres) [reply doc/source/development/rolling_operations_with_numba.rst:165]: Interesting. I think it is possible that rolling in Cython is just pretty efficient. The time I got a significant memory improvement over Pandas was comparing with a Numba EWA implementation, which is definitely more complex.

Comment (DiegoAlbertoTorres) [reply doc/source/development/rolling_operations_with_numba.rst:165]: I think it is worth mentioning somewhere in the doc that we have already found that, as N increases, the memory difference decreases and eventually Numba beats the existing impl.

File: doc/source/development/rolling_operations_with_numba.rst [hunk 7]
```diff
@@ -0,0 +1,234 @@
+Rolling Operations with Numba
```
Comment (jreback) [inline doc/source/development/rolling_operations_with_numba.rst]: Pandas Window Operations Refactor

File: doc/source/development/rolling_operations_with_numba.rst [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+~~~~~~~~~~~~~~~~~~
+
+In line with the `pandas roadmap <https://pandas.pydata.org/pandas-docs/stable/development/roadmap.html#numba-accelerated-operations>`_,
+we would like to introduce ``Numba`` as a required dependency for pandas in the 1.0 release
+and have the ``rolling.mean`` and ``rolling.apply`` operations use ``Numba`` instead of ``Cython`` in the release.
+
+Implementation
+--------------
+
+``rolling.mean`` and ``rolling.apply`` utilizes ``njit`` functions to separately calculate:
```
Comment (jreback) [inline doc/source/development/rolling_operations_with_numba.rst]: can you put a reference to njit here from the numba docs

File: doc/source/development/rolling_operations_with_numba.rst [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+
+This proof of concept (POC) demonstrates using `Numba <http://numba.pydata.org/>`_ instead of ``Cython``
+to compute ``rolling.mean`` and ``rolling.apply`` without introducing any user facing API changes.
+The benefits of using ``Numba`` include:
+
+#. Performance parity or improvement over ``Cython``
+#. Eliminate shipping C-extensions
+#. Ease of debugging and maintenance as pure Python code
+
+Additionally this POC includes an new API that allows users to create custom window boundary calculations
```
Comment (jreback) [inline doc/source/development/rolling_operations_with_numba.rst]: Add a sentence explaining *why* we would want to create custom window boundaries. this allows the user to easily overide window calculcations, for example to account for holidays or missing periods, w/o requiring changes in pandas itself.

File: doc/source/development/rolling_operations_with_numba.rst [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+when using ``Numba``.
+
+Immediate Proposal
+~~~~~~~~~~~~~~~~~~
+
+In line with the `pandas roadmap <https://pandas.pydata.org/pandas-docs/stable/development/roadmap.html#numba-accelerated-operations>`_,
+we would like to introduce ``Numba`` as a required dependency for pandas in the 1.0 release
```
Comment (jreback) [inline doc/source/development/rolling_operations_with_numba.rst]: comma after release

File: doc/source/development/rolling_operations_with_numba.rst [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+and have the ``rolling.mean`` and ``rolling.apply`` operations use ``Numba`` instead of ``Cython`` in the release.
```
Comment (jreback) [inline doc/source/development/rolling_operations_with_numba.rst]: can you reference these in the pandas docs themselves (use a full link)

File: doc/source/development/rolling_operations_with_numba.rst [hunk 12]
```diff
@@ -0,0 +1,234 @@
+Rolling Operations with Numba
+=============================
+
+Summary
+-------
+
+This proof of concept (POC) demonstrates using `Numba <http://numba.pydata.org/>`_ instead of ``Cython``
+to compute ``rolling.mean`` and ``rolling.apply`` without introducing any user facing API changes.
```
Comment (jreback) [inline doc/source/development/rolling_operations_with_numba.rst]: use the full refernce here (i left a comment about this below, but here is enough)

File: doc/source/development/rolling_operations_with_numba.rst [hunk 13]
```diff
@@ -0,0 +1,234 @@
...
+------
+
+Once ``Numba`` is a dependency in pandas, we believe it should be straightforward to implement the rest
+of the ``rolling`` aggregations, as well as ``ewa`` and ``expanding`` by using ``njit`` functions.
+
+Eventually, we aim to generalize data grouping APIs (e.g. ``rolling``, ``groupby``, ``resample``) and
+the sharing of aggregation routines (``mean``, ``apply``, ``count``) through the use of ``jitclass``.
+Currently this path is not fully explored or implemented due to performance reasons, but this issue
+will be `actively developed by the Numba team <https://github.com/numba/numba/issues/4522#issuecomment-537872456>`_
+The design document for this implementation can be found in :doc:`generalized_window`
```
Comment (jreback) [inline doc/source/development/rolling_operations_with_numba.rst]: use the full reference here (the github link)

File: doc/source/development/rolling_operations_with_numba.rst [hunk 14]
```diff
@@ -0,0 +1,234 @@
...
+
+Below is an example of creating an ``ExpandingIndexer`` that computes an expanding window similar to
+the current ``expanding`` API:
+
+.. code-block:: python
+
+   from typing import Optional
+
+   from pandas import Series
+   from pandas.core.window.indexers import BaseIndexer
```
Comment (jreback) [inline doc/source/development/rolling_operations_with_numba.rst]: create a separate issue, we may want to expose the name in pandas.api.indexers (BaseIndexer) for user consumption.

**Review by jreback — CHANGES_REQUESTED** [review]
looks good, some small changes

File: doc/source/development/rolling_operations_with_numba.rst (hunk 2 above)
Comment (mroeschke) [reply doc/source/development/rolling_operations_with_numba.rst:165]: Sounds good. Noted the long term memory pattern

**Comment by mroeschke** [comment]
Will merge this in Monday evening and write the corresponding pandas issue + mailing list if there are no further comments.

File: doc/source/development/rolling_operations_with_numba.rst [hunk 15]
```diff
@@ -0,0 +1,245 @@
...
+
+* `Window boundaries based on the window size and index <https://github.com/twosigma/pandas/blob/feature/generalized_window_operations/pandas/core/window/indexers.py>`_
+* `The actual aggregation function (mean and apply in this POC) <https://github.com/twosigma/pandas/blob/feature/generalized_window_operations/pandas/core/window/aggregators/methods.py>`_
+
+The functional implementation mimics the current ``Cython`` implementation; however, now that
+the calculation of the window boundaries is separate from the aggregation function, we are able to
+expose a new API to users that allows them to specify how to calculate the rolling window boundaries.
+
+The full implementation can be found on `this branch <https://github.com/twosigma/pandas/tree/feature/generalized_window_operations>`_
+of Two Sigma's pandas repo.
```
Comment (jreback) [inline doc/source/development/rolling_operations_with_numba.rst]: you don’t need to mention two sigma explicitly ; just point to the repo

File: doc/source/development/rolling_operations_with_numba.rst [hunk 16]
```diff
@@ -0,0 +1,245 @@
...
+       ):
+           """
+           Examples
+           --------
+           >>> ExpandingIndexer().get_window_bounds(10)
+
+           (array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0]),
+            array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10]))
+           """
+           return np.zeros(num_values, dtype=np.int64), np.arange(1, num_values + 1)
```
Comment (jreback) [inline doc/source/development/rolling_operations_with_numba.rst]: use int32

**Review by jreback — APPROVED** [review]
lgtm minor comments

**Pull request merged**